### PR TITLE
Run2-hcx147 Backport  of #19910 fixes for usage of HcalGeoemtry

### PR DESCRIPTION
**Pull request by bsunanda**
Solves Halo estimates increasing efficiency substantially

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by perrotta** [comment]
backport of #19910, with a few differences

**Comment by bsunanda** [comment]
@perrotta Please approve this back porting

**Comment by slava77** [comment]
backport of #19910
to get the label right

**Comment by slava77** [comment]
Based on https://twiki.cern.ch/twiki/bin/viewauth/CMS/MissingETOptionalFiltersRun2#Moriond_2017
the TightHalo2016 filter is supposed to be applied also on MC.
What is the inefficiency of it after the fix in MC?
At this point, we can not fix the 92X MC.

From the plots in #19910, the filter seems unusable without the fix.
It looks like this fix will not make it to the rereco done in 92X, which will limit its usefulness.
So, this is in the scope of "fix is relevant for just data and will be used with promptreco to gain experience" and is perhaps OK for 92X.

**Comment by zdemirag** [comment]
The bug will effect the efficiency in mc and data in similar ways. 

We have studied the efficiencies in the MC in 80X (no bugs because the geometry was not changed), and you can find the efficiencies on QCD and DY in the following presentation:

https://indico.cern.ch/event/576253/contributions/2331972/attachments/1353502/2044455/report_mistag_rate.pdf

slide 6 (inclusive) and slide 13 as a function of various variables.  

As you point out without the fix the filters are unusable in data. However we do have a back up both for data and mc to go to an other working point (as pointed out in the twiki you point out) which does not have the effect of this geometry. In either case, if we can include this fix in the re-reco of the data it would be good.

**Comment by slava77** [comment]
> if we can include this fix in the re-reco of the data it would be good

it's too late to push changes in the 92X rereco.
However, this probably be still added to the remaining prompt just because it changes a completely broken flag to something that could be still useful for experts.

File: RecoMET/METAlgorithms/src/HcalHaloAlgo.cc [hunk 1]
```diff
@@ -23,14 +23,11 @@ bool CompareTowers(const CaloTower* x, const CaloTower* y ){
   return x->iphi()*1000 + x->ieta() < y->iphi()*1000 + y->ieta();
 }
 
-HcalHaloAlgo::HcalHaloAlgo()
-{
+HcalHaloAlgo::HcalHaloAlgo() : geo_(nullptr), hgeo_(nullptr) {
```
Comment (slava77) [inline RecoMET/METAlgorithms/src/HcalHaloAlgo.cc]: here and in the rest of this PR please implement the backport in a way identical to changes in #19910 (unless there was something after #19910 that added new changes currently in the master).
In this file I see that at least a half of the changes are not the same in text, even though they appear to be logically the same.

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/20289	639060e
- changes in the backport are identical to https://github.com/cms-sw/cmssw/pull/19910	1bb9470
- jenkins tests pass and comparisons show changes in 2017 data workfow

This should affect both data and MC in the BeamHaloSummary flag Global\*TightHaloId2016: without the fix the flag is unusable. So, while this violates the consistency policy, the limited scope and currently unusable output appear to be enough for an exception to let this be merged.

**Comment by davidlange6** [comment]
+1

**Pull request merged**